---
title: "Johns Hopkins Summer Institute: Epigenomics in Public Health Lab 2020"
subtitle: "01 Installing and loading packages, reading in datasets"
author: "Kelly Bakulski, David Sosnowski, Shan Andrews, Brion Maher"
date: "May 29, 2020"
output: html_document
---

# Setting up R markdown format
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Installing relevant packages(only do once per computer)

```{r install packages, include=FALSE, eval=FALSE}
#We have these packages already downloaded on the cloud hence eval=FALSE. 
#If you're doing this on a personal computer you can set eval=TRUE to download there.

#Install BiocManager. This let's us install Bioconductor packages that we want to use.
if(!requireNamespace("BiocManager")) install.packages("BiocManager")

BiocManager::install(c(
  "bumphunter", #Differentially methylated region analysis
  "minfi", #Many functions for methylation analysis
  "sva", #Batch effect correction
  "limma", #Single-site association analysis
  "IlluminaHumanMethylation450kmanifest", #Manifest file
  "IlluminaHumanMethylation450kanno.ilmn12.hg19", #Annotation file
  "GO.dv", #Gene ontology queries
  "topGO",
  "missMethyl",
  "GEOquery" 
))

#Install standard R packages that we want to use.

install.packages(c(
  "RColorBrewer", #Color palettes for data visualization,
  "ggplot2", #Useful and aesthetic plotting functions
  "MatrixEQTL", #Fast eQTL and meQTL analyses 
  "qqman", #QQ and manhattan plots
  "nlme", #Non-linear mixed effects models
  "BiasedUrn" #Hypergenometric distributions
))
```

# Load relevant packages(do this every time you do an analysis)

```{r load packages}
library(minfi)
library(limma)
library(matrixStats)
library(MASS)
library(abind)
library(IlluminaHumanMethylation450kmanifest)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(sva)
library(Hmisc)
library(missmethyl)
library(ggplot2)
```

# Setting file paths for data, output, scripts

```{r source data/output}
#Setting file paths for data, output, scripts

#The rest of this script assumes that your data are in a folder called "project" on the Cloud.
#It also assumes that your ouptut will be stored in a subfolder called "output" on the Cloud. 
#As you work on your own computer, you will need to specify the folder locations.

# Folder location of the data files
#data_dir <- "/cloud/project/"
data_dir <- "~/Documents/GitHub/Application-and-Analysis-of-Epigenetic-Data-in-Public-Health-Research/data/" #Kelly's personal computer directory
data_dir
# Folder location to put the output files
#output_dir <- "/cloud/project/output/"
output_dir <- "~/Documents/GitHub/Application-and-Analysis-of-Epigenetic-Data-in-Public-Health-Research/output/" #Kelly's personal computer directory
output_dir
# Record the current date for record keeping and appending to output file names
date<-format(Sys.Date(), "%Y%m%d") 
date
```

# Read in the data

```{r load data}
pheno <- read.csv(file.path(data_dir, "samplesheet.csv"), header = TRUE, stringsAsFactors = FALSE)
dim(pheno)
head(pheno)

RGset <- read.metharray.exp(file.path(data_dir, "/idats"), targets = pheno, verbose = TRUE)
dim(RGset)

manifest <- getManifest(RGset)
str(manifest)

annotation <- getAnnotation(RGset)
dim(annotation)
annotation[1:2,]
```

# Explore the dataset

```{r explore dataset}
typeof(annotation)
typeof(RGset)
getClass(RGset)

manifest
head(getProbeInfo(manifest))
dim(getProbeInfo(manifest))
table(getProbeInfo(manifest)$Color)

pd <- pData(RGset)
table(pd$casestatus)
table(pd$gender)
table(pd$casestatus, pd$gender)
table(pd$Batch)
table(pd$casestatus, pd$Batch)
dim(pd)
length(pd$GEOID)
summary(pd$age)
summary(pd$age[ pd$gender == "M" ])
summary(pd$age[ pd$gender == "F" ])
head(pd)
```

# Save RGset object

```{r save RGset}
#Save RGChannelSet object so we can load that directly rather than loading in idats again (computationally more expensive)
save(RGset, file = file.path(data_dir, "RGset.rda"))
```
