---
title: "Epigenomics for Social Scientists"
subtitle: "01 Installing and loading packages, reading in datasets"
author: "Kelly Bakulski, Shan Andrews, John Dou, Jonah Fisher, Erin Ware"
date: "Last compiled on `r format(Sys.Date(), '%B %d, %Y')`"
output: 
  pdf_document:
  toc: true
  toc_float: true
  number_sections: true
  theme: sandstone
  highlight: tango
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = T, warning = F, message = F)
```

# Setup

## Install required packages (Do not run the following code chunk if on rstudio cloud account for EGESS)

There are many useful packages for DNA methylation pre-processing and analysis. The following packages are largely downloaded from **Bioconductor** which holds myriad useful packages for bioinformatics; the remaining packages come from CRAN which is a standard repository for general R packages. Note that your rsudio cloud environment will already have these installed so there is no need to run the following (it will take a long time). We include this code in the document so you can have a resource for package installation in future analyses you may do.

```{r install packages, include=FALSE, eval=FALSE}
# We have these packages already downloaded on the cloud hence eval=FALSE.
# If you're doing this on a personal computer you can set eval=TRUE to download there.

# Install BiocManager. This let's us install Bioconductor packages that we want to use.
if (!requireNamespace("BiocManager")) install.packages("Bioconductor")

BiocManager::install(c(
  "bumphunter", # Differentially methylated region analysis
  "minfi", # Many functions for methylation analysis
  "sva", # Batch effect correction
  "limma", # Single-site association analysis
  "IlluminaHumanMethylation450kmanifest", # Manifest file
  "IlluminaHumanMethylation450kanno.ilmn12.hg19", # Annotation file
  "GO.dv", # Gene ontology queries
  # "missMethyl",
  "GEOquery",
  "wateRmelon" #For QC and 
))

# Install standard R packages that we want to use.

install.packages(c(
  "data.table", #Helpful data formatting tools
  "ggsci", #Color palettes inspired by scientific publications and tvshows
  "viridis", #Color palette helpful for colorblindness
  "RColorBrewer", # Color palettes for data visualization,
  "tidyverse", # Useful for ggplot plotting and 
  "MatrixEQTL", # Fast eQTL and meQTL analyses
  "qqman", # QQ and manhattan plots
  "nlme", # Non-linear mixed effects models
  "BiasedUrn" # Hypergeometric distributions
))
```

## Load relevant packages

This should be done whenever you start a new r session. For this script we only use functions from the package *minfi*. This package has essentially the largest set of functions of any on Bioconductor. 

```{r load packages, message=F, warning=F}
library(minfi)
library(data.table)
library(magrittr)
library(knitr)
```

## Setting file paths for data

Here we just set up paths for loading of data

```{r source data}
# Setting file paths for data

# The rest of this script assumes that your data are in a folder called "project" on the Cloud.
# As you work on your own computer, you will need to specify the folder locations.

# Folder location of the data files
data_dir <- "E:/GESS/2060001/Data/"# for rstudio cloud switch to: "/cloud/project/Data/"
data_dir
```

# Read in the data

Here we read in our phenotype data and our RGChannelSet. The RGset is a single large object that is an amalgamation of the .idat files where the data are organized and summarised in an accessible and convenient way.

```{r load phenotype data and create rgset}
pheno <- data.table::fread(file.path(data_dir, "samplesheet.csv")) #Base R used read.csv() to read in comma seperated value files
dim(pheno)
head(pheno)

RGset <- read.metharray.exp(file.path(data_dir, "idats"), targets = pheno, verbose = TRUE, extended = T)
dim(RGset)

manifest <- getManifest(RGset)
str(manifest)

annotation <- getAnnotation(RGset)
dim(annotation)
annotation[1:4, ]
```

# Explore the dataset

```{r explore rgset, manifest, annotation, pd}
typeof(annotation)
typeof(RGset)
getClass(RGset)

manifest
head(getProbeInfo(manifest))
dim(getProbeInfo(manifest))
table(getProbeInfo(manifest)$Color)

pd <- RGset@colData@listData ; setDT(pd) #This setDT function lets us format our data as data.table

dim(pd)
head(pd)

pd[, table(casestatus)]
pd[, table(gender)]
pd[, table(gender, casestatus)]
pd[, table(Batch)]
pd[, table(Batch, casestatus)]

pd[, summary(age)]
pd[gender == "M", summary(age)]
pd[gender == "F", summary(age)]

head(pd)
```

# Save RGset object

While in our 17 sample example for lab no process takes especially long, once you scale up the number of samples you will see larger and larger increases in computation time. Therefore, saving large intermediate data products such as the RGChannelSet is helpful.

```{r save RGset}
# Save RGChannelSet object 
save(RGset, file = file.path(data_dir, "RGset.rda"))
```

