---
title: "Epigenomics for Social Scientists"
subtitle: "03 Single Site, Gene Ontology Association Analysis"
author: "Kelly Bakulski, Shan Andrews, John Dou, Jonah Fisher, Erin Ware"
date: "Last compiled on `r format(Sys.Date(), '%B %d, %Y')`"
output: 
  pdf_document:
  toc: true
  toc_float: true
  theme: sandstone
  highlight: tango
---

```{r setup, include=FALSE, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F)
```

# Setup

## Load relevant packages 

```{r load packages, message = F}
library(minfi)
library(limma)
library(matrixStats)
library(MASS)
library(sva)
library(Hmisc)
#library(missMethyl)
library(tidyverse)
library(qqman)
library(here)
library(ggsci)
library(ggpubr)
```

# Read in the data

```{r load data}
# Need noob, combat.beta, pd files
load(here("Data", "Premade_Intermediate_Files","combat-beta.rda"))
load(here("Data", "pdqc.rda"))

# Make sure this outputs to TRUE
all.equal.character(pd$Basename, colnames(combat.beta))
```

# Principal components on cell proportions

```{r principal components on cell proportions}
celltypes <- c("GR","NK","B","CD4","CD8","MO")
cellpcs <- prcomp(t(pd[celltypes]))
pcvar <- data.frame(pc = seq(1, length(cellpcs$sdev)), var = cellpcs$sdev^2 / sum(cellpcs$sdev^2))

ggplot(pcvar, aes(x = pc, y = var)) +
  geom_bar(stat = "identity") +
  labs(x = "Variance Explained", y = "Principal Component") +
  theme_bw()

summary(pd$GR)
```

# Box Plots and t-statistics for each cell type

```{r Box Plots and t-statistics for each cell type}
# Box Plots and t-statistics for each cell type
celldf <- pivot_longer(pd[c(celltypes, "casestatus")], cols = celltypes)

ggplot(celldf, aes(x = name, y = value, fill = casestatus)) +
  geom_boxplot() +
  labs(x = "Cell type", y = "Proportion", fill = "Case status") +
  theme_bw() + 
  scale_fill_nejm() +
  stat_compare_means(label = c("p.format"), method = "t.test", vjust = 20) +
  stat_compare_means(label = c("p.signif"), method = "t.test", vjust = 18) 
```

# Single site association testing

```{r single site association testing}
# Construct the model matrix
mod <- model.matrix(~ factor(pd$casestatus) + pd$age + factor(pd$sex) + factor(pd$smoking) + pd$GR)

# Run the single site association model
out <- lmFit(combat.beta, mod)
out <- eBayes(out)
ss.hits <- topTable(out, coef = 2, number = nrow(combat.beta))

head(ss.hits, n = 10)

rm(mod, out)
```

## QQ plot with lambda statistic

```{r qq plot}
observed <- -log10(sort(ss.hits$P.Value, decreasing = F))
expected <- -log10(ppoints(length(ss.hits$P.Value)))
lambda <- median(observed) / median(expected)

# Make a qq plot of our data
qq(ss.hits$P.Value, main = sprintf("Lambda value of %.3f", lambda))
```

## Make a manhattan plot of our data

Manhattan plots place negative log10 transformed P values on the y axis. After the transformation, a higher value indicates greater significance. The X axis is ordered by chromosome number and position within the genome. 

Manhattan plots are not quite as helpful with Epigenetics as they are with Genetics. This is partly due to how significant locations differ between DNA and DNAm. While genetic hits often cluster within GWAS, epigenetic hits from EWAS tend to not cluser together very much.

Nonetheless, it may still be useful to take a look at one.

```{r prepare for manhattan plot}
anno <- IlluminaHumanMethylation450kanno.ilmn12.hg19::Locations
anno$chr <- str_split(anno$chr, "chr",n=2,simplify=T)[,2]
anno <- anno[anno$chr %in% 1:22,]
anno$chr <- as.numeric(anno$chr)

forman <- merge(anno[c("chr", "pos")], ss.hits[c("P.Value")], by = "row.names")
forman <- forman[complete.cases(forman),]
forman$SNP <- ""
```

```{r manhattan plot}
# Call function
qqman::manhattan(forman, chr = "chr", bp = "pos", p = "P.Value")
```

## Volcano Plot

Volcano plots show effect size estimates (positive and negative) on the x axis against log transformed P values on the y axis. Typically the higher the absolute value of the effect size, the more significant the p value is. This results in a distribution of points that resembles a volcano. 

```{r volcano plot}
ggplot(ss.hits, aes(x = logFC, y = -log10(P.Value))) +
  geom_point() +
  theme_bw() +
  labs(x = "Log Fold Change", y = "-log10 P Value")
```


## Plot the top 6 hits and examine case vs control beta differences

```{r plot single site top hits}
# Create a mini result list of the first 6 hits
tophits <- ss.hits[1:6, ]

# Function to make plot of one CpG
plotop <- function(hit){
  dat <- data.frame(vals = combat.beta[hit,], case = pd$casestatus)
  mdiff <- mean(dat$vals[dat$case == "RA"]) - mean(dat$vals[dat$case == "Control"])
  ggplot(dat, aes(x = case, y = vals, color = case)) +
    geom_boxplot(outliers = F, fill = "grey", width = 0.5) +
    geom_jitter(width = 0.1) +
    labs(x = "Case Status", y = "Beta Methylation Value", title = sprintf("Probe %s", hit),
         subtitle = sprintf("FDR P = %.3f. Mean Diff = %.3f", tophits$adj.P.Val[rownames(tophits) == hit], mdiff),
         color = "Case Status") +
    scale_color_nejm() +
    scale_y_continuous(limits = c(0, 1), breaks = seq(0, 1, by = 0.25), labels = c("0", "25", "50", "75", "100")) +
    theme_bw()
}

# Version without y-axis set to 0-100% methylation
# plotop <- function(hit){
#   dat <- data.frame(vals = combat.beta[hit,], case = pd$casestatus)
#   mdiff <- mean(dat$vals[dat$case == "RA"]) - mean(dat$vals[dat$case == "Control"])
#   ggplot(dat, aes(x = case, y = vals, color = case)) +
#     geom_boxplot(outliers = F, fill = "grey", width = 0.5) +
#     geom_jitter(width = 0.1) +
#     labs(x = "Case Status", y = "Beta Methylation Value", title = sprintf("Probe %s", hit),
#          subtitle = sprintf("FDR P = %.3f. Mean Diff = %.3f", tophits$adj.P.Val[rownames(tophits) == hit], mdiff),
#          color = "Case Status") +
#     scale_color_nejm() +
#     theme_bw()
# }


# Loop through top six CpGs and use plot function
for(cpg in rownames(tophits)){
  tophit_plot <- plotop(cpg)
  print(tophit_plot)
}
rm(formanhattan)
```

# Gene ontology analysis

```{r calculate gene ontology enrichment}
# Need to pick a significance cutoff for inclusion. Here is 1x10-5, but may need to be flexible with data.
#gene.ontology <- missMethyl::gometh(as.character(rownames(ss.hits)[ss.hits$P.Value < 1e-5]), all.cpg = as.character(rownames(ss.hits)), plot.bias = FALSE, prior.prob = TRUE)
#save(gene.ontology, file = here("Data", "Premade_Intermediate_Files", "gene-ontology.rda"))
# memory requirements of missMethyl package may be too much for RStudio cloud, load the premade gene.ontology object:

load(here("Data", "Premade_Intermediate_Files","gene-ontology.rda"))

dim(gene.ontology)
summary(gene.ontology$P.DE)
summary(gene.ontology$FDR)
gene.ontology <- gene.ontology[order(gene.ontology$P.DE), ]
head(gene.ontology)
```

```{r plot gene ontology}
# Some functions that will be useful in making the plot
wrap.it <- function(x, len) {
  sapply(x, function(y) {
    paste(strwrap(y, len),
      collapse = "\n"
    )
  },
  USE.NAMES = FALSE
  )
}
# Call this function with a list or vector
wrap.labels <- function(x, len) {
  if (is.list(x)) {
    lapply(x, wrap.it, len)
  } else {
    wrap.it(x, len)
  }
}

par(mai = c(1, 4, 1, 1))
barplot(abs(log(as.numeric(gene.ontology$P.DE[1:10]), base = 10)),
  main = "Gene Ontology", horiz = TRUE, names.arg = wrap.labels(gene.ontology$TERM[1:10], 50),
  xlab = "-log10(P value)", col = "dodgerblue", las = 1, cex.axis = 1.2, cex.main = 1.4, cex.lab = 1, cex.names = 1, space = 0.4, xlim = c(0, 5)
)
```
